[19766][Refactoring] Toolbar Helper

floriank opened the pull request:
## Toolbar component

We introduced a toolbar component with `release/4.1` which simplifies the generation of the content header for all of the individual sections of OP. the component came with a helper that made code generation a little easier:

``` erb
<%= toolbar title: l(:label_overview) do %>
  <% if User.current.allowed_to?(:add_subprojects, project) %>
    <li class="toolbar-item">
      <%= link_to new_project_path(parent_id: project), class: 'button -alt-highlight' do %>
        <i class="icon-add"></i> <%= l(:label_subproject_new) %>
      <% end %>
    </li>
  <% end %>
  <% if User.current.allowed_to?(:edit_project, project) %>
    <li class="toolbar-item">
      <%= link_to my_projects_overview_path(project), class: 'button', accesskey: accesskey(:edit) do %>
        <i class="icon-edit"></i> <%= l(:label_personalize_page) %>
      <% end %>
    </li>
  <% end %>
<% end %>
```

While being better than using plain HTML, this is still hard to read as the view is cluttered.
## What this PR does

The PR introduces a second helper in addition to the `ToolbarHelper` - as correctly pointed out by @NobodysNightmare, the toolbar is not the component, the content header is. the new helper is thereby called `ContentHeaderHelper` and provides a DSL for creating the header - the code above can be simplified as:

``` erb
<%= content_header title: l(:label_overview) do |header| %>
  <%= header.toolbar do |toolbar| %>
    <%= toolbar.button l(:label_subproject_new), new_project_path(parent_id: project), icon: :add, highlight: :alt, if: User.current.allowed_to?(:add_subprojects, project) %>
    <%= toolbar.button l(:label_personalize_page), my_projects_overview_path(project), icon: :edit, accesskey: :edit, if:  User.current.allowed_to?(:edit_project, project) %>
  <% end %>
<% end %>
```

The DSL provides some simplifications such as `icon` and `accesskey` options to the buttons directly and detached the DOM necessary for the component from the individual view, DRYing up all of the partials nicely.
### Additionals

This PR also cleans up some of the path helpers and links, removes a little bit of dead code
## Still TODO
- [x] Watcher Links for the toolbar
- [x] Submenus in the toolbar (and the replacement on wiki and work package Legacy pages)
- [x] full removal of the usage of the `_action_menu_*` partials
- [ ] Tests for the `ContentHeaderHelper` module and contained classes
## Plugins?

The old `ToolbarHelper` component stays in the code base until all plugins have been updated properly.

Commit:
introduce a new content header helper and test it out on the users

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
introduce a new content header helper and test it out on the users

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
restore hook and use proper button in users index

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
restore hook and use proper button in users index

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
refactor the content header helper

also, delegate accesskeys to OpenProject::AccessKerys

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
refactor the content header helper

also, delegate accesskeys to OpenProject::AccessKerys

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header helper in user administration views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header helper in user administration views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in groups administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in groups administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in role administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in role administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in types and project types

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in types and project types

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar button respect link options

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar button respect link options

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header helper in wp status administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header helper in wp status administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in workflows

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in workflows

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in custom field administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in custom field administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in enumerations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in enumerations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in settings adminsitration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in settings adminsitration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in auth sources administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in auth sources administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in colour administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in colour administration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in info section

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content_header in info section

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use conten_header in timelines

also, clean up the timelines helper and introduce a named route for destruction
confirmation

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use conten_header in timelines

also, clean up the timelines helper and introduce a named route for destruction
confirmation

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in work package views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in work package views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in account views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in account views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in activity index

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in activity index

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use conten header in plugin and project administrations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use conten header in plugin and project administrations

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in attachments

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in attachments

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in board views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in board views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in category adminsitration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in category adminsitration

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header for copy projects

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header for copy projects

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in messages views

also provides some further cleanup in the views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in messages views

also provides some further cleanup in the views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in the core my page views

(no this does not update the my project page plugin)

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in the core my page views

(no this does not update the my project page plugin)

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in news views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in news views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in project management

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in project management

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in reportings

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in reportings

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in repository views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in repository views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in seach view

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in seach view

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in time entries

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in time entries

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in timelogs

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in timelogs

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in version views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in version views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in compatible wiki views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in compatible wiki views

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use missing content header in colour administration

forgot that one, heh

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use missing content header in colour administration

forgot that one, heh

Signed-off-by: Florian Kraft <f.kraft@finn.de>

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://floriank-refactoring-toolbar-helper.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://floriank-refactoring-toolbar-helper.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

Commit:
silence the hound

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
silence the hound

[ci skip]

Signed-off-by: Florian Kraft <f.kraft@finn.de>

NobodysNightmare commented:
I'd recommend marking the old `ToolbarHelper` with deprecation warnings. As far as I could see you have not done that yet.

floriank commented:
You're correct, I have not, the essential plan is to switch the plugins as soon as possible as well

NobodysNightmare commented:
_essential plan_™

Commit:
add basic submenu functionality in toolbar

also, document it and provide some misc javascript to build the menu.

This _knowingly_ duplicates existing functionality as the current implementation
for submenus on the WP List is too specific to the angular module itself and will
be extremely inconvenient to extract.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add basic submenu functionality in toolbar

also, document it and provide some misc javascript to build the menu.

This _knowingly_ duplicates existing functionality as the current implementation
for submenus on the WP List is too specific to the angular module itself and will
be extremely inconvenient to extract.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair a conflict with the work package list

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair a conflict with the work package list

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
switch to classes

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
switch to classes

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use the content header in work package show (legacy)

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use the content header in work package show (legacy)

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar submenu helpers work

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar submenu helpers work

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
provide the work packages "more" menu again

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
provide the work packages "more" menu again

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
reorder buttons in work pacakges more menu

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
reorder buttons in work pacakges more menu

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
provide a watch button for the toolbar component

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
provide a watch button for the toolbar component

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar scrollable

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the toolbar scrollable

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a little bit of fine tuning for the wp legacy form

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add a little bit of fine tuning for the wp legacy form

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
move available format links to to toolbar

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
move available format links to to toolbar

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank committed:
user new watcher link implementation in core functionality

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
user new watcher link implementation in core functionality

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
WIP

[ci skip]

floriank committed:
WIP

[ci skip]

Commit:
use new toolbar on wiki show page

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use new toolbar on wiki show page

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in wiki ToC

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in wiki ToC

Signed-off-by: Florian Kraft <f.kraft@finn.de>

ulferts commented on app/helpers/content_header_helper.rb:
```diff
@@ -0,0 +1,211 @@
+
+    def items(&block)
+      return @helper.capture(ContentToolbar.new, &block) if block_given?
+      ''
```
@floriank isn't this is weird way of writing

```
if block_given?
  @helper.capture(ContentToolbar.new, &block)
else
  ''
end
```

NobodysNightmare replied:
@ulferts :+1:

floriank replied:
Yeah, but it is 2 lines instead of 5. 

I could change it to:

``` ruby
return '' unless block_given?
@helper.capture(ContentToolbar.new, &block)
```

NobodysNightmare replied:
Readability is not always about line counting ^^

As I generally like guard clauses I would still agree to your proposal.

floriank replied:
I will do the refactoring once the new features have tests to verify functionality.

I concur on the topic of guard clauses - however, i really dislike `if ... else` statements, in part because i like to aim for methods under 10 lines. Wasn't there a rubocop rule for that :wink: ?

ulferts commented on app/controllers/watchers_controller.rb:
```diff
@@ -105,20 +105,7 @@ def set_watcher(user, watching)
-
-        render action: 'replace_selectors'
-      end
+      format.js
```
If you remove the `@replace_selectors` variable here, it should also be removed in the rjs template.

floriank replied:
the `rjs` temlate is gone, the current implementation uses `watch.js` and `unwatch.js` (as defined by Rails magic) - I am currently looking into merging these two.

ulferts commented on app/helpers/timelines_helper.rb:
```diff
@@ -183,35 +183,6 @@ def timeline_options
-            class: 'button',
-            &block
-            )
-  end
```
:+1:

Commit:
add keybindings to the toolbar

also add generated tabindex in order of appearance

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add keybindings to the toolbar

also add generated tabindex in order of appearance

Signed-off-by: Florian Kraft <f.kraft@finn.de>

ulferts commented on app/helpers/content_header_helper.rb:
```diff
@@ -0,0 +1,211 @@
+        end
+      else
+        content_tag :li, '', options
+      end
```
```
content_tag :li, options do
  if block_given?
    yield
  else
    ''
  end
end

```

Together with my other comment regarding returning '' on block_given? I see a pattern emerging here.

floriank replied:
And that's why we have the WIP label ;)

Commit:
improve javascript style

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
improve javascript style

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add the toolbar to possible action menus in the menu steps

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add the toolbar to possible action menus in the menu steps

Signed-off-by: Florian Kraft <f.kraft@finn.de>

ulferts commented:
@floriank I like the effort you put into this in order to have a unified interface for the menus.

I haven't yet let the whole thing run on my machine but from code review I noticed:
- the inclusion of watch_button in the ContentToolbar class is breaking the class' abstraction but as of now I haven't yet found a better way of organising the code without complicating the interface.
- There are no tests for the ContentHeaderHelper. If we want to later refactor the code, they would be extremely helpful.

floriank commented:
@ulferts see todo list above. Tests will be included ~after~ I have fixed the currently breaking specs.

As for the watch button - I am not sure i follow here - the `watch` button is a new implementation on purpose as the old implementation is quite inflexible. 

Ideally it should be just another button handling a route called via Rails' `remote: true`, but there is a little bit of text/method/option replacement magic that has to happen.

ulferts commented:
@floriank don't get me wrong on the watch button. I like for it to be rewritten as the old implementation wasn't that great. What I dislike is the signature 'watch_button' and the resulting code within the `ContentHeaderHelper` module.

Commit:
make some cukes pass again by looking for the toolbar markup

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make some cukes pass again by looking for the toolbar markup

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair timelines spec and use content_header in reportings

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair timelines spec and use content_header in reportings

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
group export functions in toolbars for wiki and work packages

also, truly hide the toolbar's submenus

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
group export functions in toolbars for wiki and work packages

also, truly hide the toolbar's submenus

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the buttons in the toolbar properly clickable

should also take care of some capybara related clicking issues

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
make the buttons in the toolbar properly clickable

should also take care of some capybara related clicking issues

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in versions

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
use content header in versions

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair cucumber tests for work package

somehow this only works with waits

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
repair cucumber tests for work package

somehow this only works with waits

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
compensate for select2 escaping the script tags

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
compensate for select2 escaping the script tags

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
fix timelines controller spec

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
fix timelines controller spec

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
switch to xpath for looking for the copy link

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
switch to xpath for looking for the copy link

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
fix the news page when no project is given

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
fix the news page when no project is given

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
hide new user button when password login is disabled

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
hide new user button when password login is disabled

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank commented:
Note: This PR somewhat deals with https://community.openproject.org/work_packages/19766

Commit:
remove replace selectors partial and clean up watch tests

the replace multiple selectors has no function anymore as the watchers tab
will now be rendered independently of the watch/unwatch call.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove replace selectors partial and clean up watch tests

the replace multiple selectors has no function anymore as the watchers tab
will now be rendered independently of the watch/unwatch call.

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
check if the user may see the watchers in the watch/unwatch cycle

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
check if the user may see the watchers in the watch/unwatch cycle

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
update date_index action with content_header

also, fix the breaking tests along the way

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
update date_index action with content_header

also, fix the breaking tests along the way

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add missing content_header uses

also, cleanup a little

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
remove empty file

Commit:
do not use scrollable toolbar on modal layout for wp

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add copyright header

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
do not allow empty classes on links in the toolbar

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
add dummy specs

Signed-off-by: Florian Kraft <f.kraft@finn.de>

Commit:
WIP components for frontend

[ci skip]

Commit:
Merge branch 'refactoring/toolbar-helper' of github.com:floriank/openproject into refactoring/toolbar-helper

Conflicts:
  app/helpers/content_header_helper.rb
  app/views/wiki/show.html.erb
  app/views/work_packages/show.html.erb

Commit:
provide an initial DSL that uses the ui components

Signed-off-by: Florian Kraft <f.kraft@finn.de>

floriank committed:
add a watch button and make all the dsl calls conditionable

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
apply includes for DSL

unfortunately, as we instance_eval a lot, the DSL ahs to know about stuff
like Rails routes and Redmine::I18n

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
apply new DSL in admi projects

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
cleanup for the hound

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank committed:
remove extra line

[ci skip]

Signed-off-by: Florian Kraft <FlorianKraft@gmx.de>

floriank commented:
The topic is on hold till OP 4.3 - I will create another PR at a later point in time and close this one for now.
